### ISSUE TITLE
Traceback on duplicate entries in staticpage agreements

### ISSUE THREAD

**Issue by dwaynebailey** [translate/pootle#3815]
```pytb
 > staticpages:0012_data__adjust_pks
 - Migration 'staticpages:0012_data__adjust_pks' is marked for no-dry-run.2015-06-10 17:36:41,366 WARNING /home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/dj
ango/db/backends/mysql/base.py:124: Warning: Column 'user_id' cannot be null
  return self.cursor.execute(query, args)
FATAL ERROR - The following SQL query failed: ALTER TABLE `staticpages_agreement` ADD CONSTRAINT `staticpages_agreement_user_id_26f0568b2f8d7f1e_uniq` UNIQUE (`user_id`, `document_id`)The error was: (1062, "Duplicate entry '0-1' for key 'staticpages_agreement_user_id_26f0568b2f8d7f1e_
uniq'")
 ! Error found during real run of migration! Aborting.

 ! Since you have a database that does not support running
 ! schema-altering statements in transactions, we have had 
 ! to leave it in an interim state between migrations.
! You *might* be able to recover with:   (migration cannot be dry-run; cannot discover commands)
 ! The South developers regret this has happened, and would
 ! like to gently persuade you to consider a slightly
 ! easier-to-deal-with DBMS (one that supports DDL transactions)
 ! NOTE: The error which caused the migration to fail is further up.
Error in migration: staticpages:0012_data__adjust_pks
Traceback (most recent call last):
  File "./manage.py", line 32, in <module>
    execute_from_command_line()
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/django/core/management/__i
nit__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/django/core/management/__i
nit__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/django/core/management/bas
e.py", line 242, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/django/core/management/bas
e.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/management/[35/1153]
migrate.py", line 111, in handle
    ignore_ghosts = ignore_ghosts,
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/migration/__init__.p
y", line 220, in migrate_app
    success = migrator.migrate_many(target, workplan, database)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/migration/migrators.
py", line 256, in migrate_many
    result = migrator.__class__.migrate_many(migrator, target, migrations, database)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/migration/migrators.
py", line 331, in migrate_many
    result = self.migrate(migration, database)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/migration/migrators.
py", line 133, in migrate
    result = self.run(migration, database)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/migration/migrators.
py", line 114, in run
    return self.run_migration(migration, database)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/migration/migrators.
py", line 84, in run_migration
    migration_function()
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/migration/migrators.
py", line 60, in <lambda>
    return (lambda: direction(orm))
  File "/home/dwayne/pootle/pootle/apps/staticpages/migrations/0012_data__adjust_pks.py", line 42, in
 forwards
    db.create_unique('staticpages_agreement', ['user_id', 'document_id'])
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/db/generic.py", line
 47, in _cache_clear
    return func(self, table, *args, **opts)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/db/generic.py", line
 623, in create_unique
    cols,
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/south/db/generic.py", line
 282, in execute
    cursor.execute(sql, params)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/django/db/backends/util.py
", line 53, in execute
    return self.cursor.execute(sql, params)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/django/db/utils.py", line 
99, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/django/db/backends/util.py
", line 53, in execute
    return self.cursor.execute(sql, params)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/django/db/backends/mysql/b
ase.py", line 124, in execute
    return self.cursor.execute(query, args)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/MySQLdb/cursors.py", line 
205, in execute
    self.errorhandler(self, exc, value)
  File "/home/dwayne/.virtualenvs/pootle/local/lib/python2.7/site-packages/MySQLdb/connections.py", l
ine 36, in defaulterrorhandler
    raise errorclass, errorvalue
django.db.utils.IntegrityError: (1062, "Duplicate entry '0-1' for key 'staticpages_agreement_user_id_
26f0568b2f8d7f1e_uniq'")
```

**Comment by dwaynebailey** [comment]
I guess the right correction here is that before we do these migrations we should:

1. Delete duplicates (user would need to agree again), or
2. Take the most recent agreement acceptance (user would see nothing and we're just removing the redundant duplicate)

The second is nicer than the first but the impacted users is by my guess going to be tiny.

**Comment by jleclanche** [comment]
Duplicates shouldn't happen. That migration moves one unique to another...

**Comment by jleclanche** [comment]
Disregard that.

**Comment by jleclanche** [comment]
What results do you get with 225b1845f3420b54e36bc5076e566a4b199cc6ee ?

**Comment by unho** [comment]
Assuming that combinations of `document` and `user` (profile before the migration) are unique, and given that `profile` has a OneToOne with user (https://github.com/translate/pootle/commit/1fdce82148642cad937fe400d1df06818db60ae3#diff-dc30770eeded9edff202b86a72355ea7L33) this shouldn't happen.

'0-1' is a weird key for a user, isn't it?

**Comment by unho** [comment]
Merged a fix 3fb6aa0e. Since we couldn't trace the root cause of the above traceback we can't tell for sure if this fix solves the issue.

**Comment by jleclanche** [comment]
That would definitely be it.

**Comment by dwaynebailey** [comment]
Reopening.  So we have two issues exposed here.

1. The migration was just rubbish. Period. - that is fixed in 3fb6aa0
2. Testing on fake 2.5.1 + critical LO data is still creating a number of ``0-1`` pairs meaning that they cannot possibly be unique and causing this traceback.  So still and issue.

**Comment by jleclanche** [comment]
0-1 is an odd pair to create. What do the IDs refer to?

**Comment by dwaynebailey** [comment]
@jleclanche user_id + document_id.  My guess is we have documents with an agreement with a user that doesn't exist anymore.  I'm guessing though

**Comment by jleclanche** [comment]
Maybe a `DELETE FROM staticpages_agreement WHERE user_id = NULL` before the UPDATE would fix it.

**Comment by jleclanche** [comment]
This is fixed by removing the migration which was never actually needed in the first place. *sigh*

**Comment by dwaynebailey** [comment]
Fixed 71610bb0072c9d6b24ca9a982b5b64d89b644089 with some explanation.
